AllTalk TTS Initial commit

erew123 opened the pull request:
AllTalk is a TTS engine based on Coqui https://github.com/erew123/alltalk_tts I'm submitting a separate PR for the documentation after this PR. 

I messaged @Cohee1207 about submitting a PR for AllTalk on Discord (10th Jan).

Due to AllTalk having to the option built into it to split text further and have a narrator/character style conversation e.g. https://vocaroo.com/18nrv7FR6wuA I have added to **index.js** **"Pass Asterisks to TTS Engine"** as an option within the TTS extensions page. This setting is disabled as standard and also set to save as part of the page settings for SillyTavern. It literally does exactly what it says, allows the asterisks to not be filtered and pass them onto the TTS engine. It does not affect default behaviour of the SillyTavern TTS extensions page, so wont impact any other TTS engine.

Its quite a comprehensive TTS addon, so feel free to ask me any questions.

![image](https://github.com/SillyTavern/SillyTavern/assets/35898566/00a0b06d-7fbd-4087-8811-943b32f5384e)

Thanks

erew123

erew123 committed:
AllTalk initial commit

AllTalk TTS initial commit from https://github.com/erew123/alltalk_tts
ST index.js is updated to have "Pass Asterisks to TTS Engine" which allows AllTalk to split text as character/narrator voices. This setting has been left disabled as standard, so wont affect other TTS engines. The setting will save along with other settings on the page (if checked).

erew123 committed:
Merge branch 'SillyTavern:release' into release

Cohee1207 commented on public/scripts/extensions/tts/alltalk.js:
```diff
@@ -0,0 +1,938 @@
+export { AllTalkTtsProvider };
+
+const css = `
+.settings-row {
```
It will be better to put these styles into `public\scripts\extensions\tts\style.css`

erew123 replied:
Sorted this bit.

Cohee1207 commented on public/scripts/extensions/tts/alltalk.js:
```diff
@@ -0,0 +1,938 @@
+        try {
+            if (this.settings.at_generation_method === 'streaming_enabled') {
+                // For streaming method
+                const streamingUrl = `${this.settings.provider_endpoint}/api/tts-generate-streaming?text=${encodeURIComponent(inputText)}&voice=${encodeURIComponent(voiceId)}.wav&language=${encodeURIComponent(this.settings.language)}&output_file=stream_output.wav`;
```
You can output a url as a string from the generateTts, it will be correctly played by ST (I actually took this approach from AllTalk initially). Having a separate audio element will disconnect it from global TTS controls, as it will be impossible to stop a currently played audio when switching chats for example.

erew123 replied:
@Cohee1207 Ive tried this bit of code:

```
            if (this.settings.at_generation_method === 'streaming_enabled') {
                // For streaming method
                const streamingUrl = `${this.settings.provider_endpoint}/api/tts-generate-streaming?text=${encodeURIComponent(inputText)}&voice=${encodeURIComponent(voiceId)}.wav&language=${encodeURIComponent(this.settings.language)}&output_file=stream_output.wav`;
                console.log("Streaming URL:", streamingUrl);
    
                // Fetch the audio stream and return the response object
                const response = await fetch(streamingUrl);
                if (!response.ok) {
                    throw new Error(`HTTP ${response.status}: Failed to fetch audio stream`);
                }
                return response; 
            } else {
```

ST does play back, but not until generation is completed. In other words, its not actually streaming the audio, its waiting for a complete wav/blob to appear (as far as I can tell). Do you have any other suggestion on this? As Streaming would be a bit pointless otherwise. Im happy for any suggestions/help you can offer with this.

Cohee1207 replied:
You don't have to fetch the streaming URL, just return it from the function, TTS plugin will handle it. I added the related code here for XTTS provider streaming, look for reference: #1623

erew123 replied:
It wont handle a response as a url e.g. Ive tried this way:

```
    if (this.settings.at_generation_method === 'streaming_enabled') {
        // Construct the streaming URL
        const streamingUrl = `${this.settings.provider_endpoint}/api/tts-generate-streaming?text=${encodeURIComponent(inputText)}&voice=${encodeURIComponent(voiceId)}.wav&language=${encodeURIComponent(this.settings.language)}&output_file=stream_output.wav`;
        console.log("Streaming URL:", streamingUrl);

        // Return the streaming URL directly
        return streamingUrl; 
```

and that results in:

```
Streaming URL: http://localhost:7851/api/tts-generate-streaming?text=Hmm...%20So%2C%20you%20believe%20someone%20close%20to%20you%20may%20have%20committed%20murder%3F%20And%20you%20wish%20for%20me%20to%20investigate%20the%20matter%20without%20their%20knowledge%3F&voice=Alan_Rickman.wav&language=en&output_file=stream_output.wav
index.js:455 Current TTS job for Sherlock Holmes completed.
index.js:422 Uncaught (in promise) TypeError: response.blob is not a function
    at addAudioJob (index.js:422:38)
    at tts (index.js:474:5)
```

Though I can alter index.js so that it can differentiate e.g. 

```
async function playAudioData(audioData) {
    if (currentAudioJob == null) {
        console.log('Cancelled TTS playback because currentAudioJob was null');
        return;
    }

    if (audioData instanceof Blob) {
        // Handle Blob object (existing logic)
        const reader = new FileReader();
        reader.onload = function (e) {
            const srcUrl = e.target.result;
            audioElement.src = srcUrl;
            audioElement.addEventListener('ended', completeCurrentAudioJob);
            audioElement.addEventListener('canplay', () => {
                console.debug('Starting TTS playback');
                audioElement.play();
            });
        };
        reader.readAsDataURL(audioData);
    } else if (typeof audioData === 'string') {
        // Handle string as a URL for streaming audio
        audioElement.src = audioData;
        audioElement.addEventListener('ended', completeCurrentAudioJob);
        audioElement.addEventListener('canplay', () => {
            console.debug('Starting TTS playback');
            audioElement.play();
        });
    } else {
        console.error(`playAudioData received invalid data type ${typeof audioData}`);
    }
}
```

and

```
async function addAudioJob(responseOrUrl) {
    if (typeof responseOrUrl === 'string') {
        // If the input is a string, assume it is a URL for streaming audio
        audioJobQueue.push(responseOrUrl);
        console.debug('Pushed streaming audio URL to queue.');
    } else if (responseOrUrl instanceof Response) {
        // Handle the response object as before
        const audioData = await responseOrUrl.blob();
        if (!audioData.type.startsWith('audio/')) {
            throw `TTS received HTTP response with invalid data format. Expecting audio/*, got ${audioData.type}`;
        }
        audioJobQueue.push(audioData);
        console.debug('Pushed audio job to queue.');
    } else {
        // Handle unexpected input types
        throw `TTS received invalid audio data type ${typeof responseOrUrl}`;
    }
}
```
Would you be ok with that change?

Cohee1207 replied:
It should be already doing so, as seen here (release branch): https://github.com/SillyTavern/SillyTavern/blob/e3ccaf70a10b862113f9bad8ae039fc7ce6570df/public/scripts/extensions/tts/index.js#L322

Try pulling the latest release/staging in your branch.

erew123 replied:
I think I know exactly what happened here......I must have taken a copy of index.js literally the day or so before you made that change... Obviously Ive made other changes to index.js... so Ive been re-using that one, even though Ive updated ST to latest a few times. So thats why its not been working. I just didnt know you had made that change and Ive been re-using the old index.js. Problem solved it seems.

Cohee1207 commented on public/scripts/extensions/tts/alltalk.js:
```diff
@@ -0,0 +1,938 @@
+                    throw new Error(`HTTP ${audioResponse.status}: Failed to fetch audio data`);
+                }
+                const audioBlob = await audioResponse.blob();
+                if (!audioBlob.type.startsWith('audio/')) {
```
There's no need to duplicate these checks here as they are done on the TTS plugin level, just return a fetch Response after checking for `ok` result.

erew123 replied:
done.

Cohee1207 commented on public/scripts/extensions/tts/alltalk.js:
```diff
@@ -0,0 +1,938 @@
+function saveAllTalkSettings(settingsToSave) {
+    // Save the specific settings to localStorage
+    console.log("settings", settingsToSave)
+    localStorage.setItem('AllTalkSettings', JSON.stringify(settingsToSave));
```
Please avoid using localStorage settings as they bind them to a specific browser window. See how other TTS providers handle the settings (writing them to extension_settings).

Cohee1207 committed:
Merge pull request #1701 from SillyTavern/staging

Staging

erew123 committed:
Merge branch 'SillyTavern:release' into release

erew123 committed:
AllTalk Updates

Streaming passed URL to global ST audio.
Localstorage removed for saving TTS elements.
Styles stored in CSS
Duplicate checks on fetchresponse removed.

erew123 commented:
@Cohee1207 As far as I am aware, those are the 4x requested changes done.

I've updated my fork and also updated my 3x files on there (alltalk.js, indesx.js and style.css).

Excuse me I'm still a little new to this.... I DONT have to submit another PR now... correct? If you merge, you will pull my new changes (is what I believe after a little research).

Thanks for your time/help though and apologies for the confusion with the streamingurl/index.js (I did actually originally try that method back when I started, so you must have literally updated it the day or so after I started on it).

Any problems, please let me know

erew123

Cohee1207 committed:
Merge branch 'staging' into erew123/release

Cohee1207 committed:
Fix format
